op-node/rollup/sync/start.go

```diff
@@ -41,163 +28,197 @@ import (
+		return nil, fmt.Errorf("failed to find the finalized L2 block: %w", err)
+	}
+
+	safe, err := l2.L2BlockRefByLabel(ctx, eth.Safe)
+	if errors.Is(err, ethereum.NotFound) {
+		safe = finalized
+	} else if err != nil {
+		return nil, fmt.Errorf("failed to find the safe L2 block: %w", err)
 	}
+
+	unsafe, err := l2.L2BlockRefByLabel(ctx, eth.Unsafe)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find the L2 head block: %w", err)
+	}
+	return &FindHeadsResult{
+		Unsafe:    unsafe,
+		Safe:      safe,
+		Finalized: finalized,
+	}, nil
 }
 
-// FindL2Heads walks back from `start` (the previous unsafe L2 block) and finds the unsafe and safe
-// L2 blocks.
+// FindL2Heads walks back from `start` (the previous unsafe L2 block) and finds
+// the finalized, unsafe and safe L2 blocks.
 //
-//   - The *unsafe L2 block*: This is the highest L2 block whose L1 origin is a plausible (1)
+//   - The *unsafe L2 block*: This is the highest L2 block whose L1 origin is a *plausible*
 //     extension of the canonical L1 chain (as known to the op-node).
 //   - The *safe L2 block*: This is the highest L2 block whose epoch's sequencing window is
 //     complete within the canonical L1 chain (as known to the op-node).
+//   - The *finalized L2 block*: This is the L2 block which is known to be fully derived from
+//     finalized L1 block data.
 //
-// (1) Plausible meaning that the blockhash of the L2 block's L1 origin (as reported in the L1
-//
-//	Attributes deposit within the L2 block) is not canonical at another height in the L1 chain,
-//	and the same holds for all its ancestors.
-func FindL2Heads(ctx context.Context, start eth.L2BlockRef, seqWindowSize uint64,
-	l1 L1Chain, l2 L2Chain, genesis *rollup.Genesis) (unsafe eth.L2BlockRef, safe eth.L2BlockRef, err error) {
+// Plausible: meaning that the blockhash of the L2 block's L1 origin
+// (as reported in the L1 Attributes deposit within the L2 block) is not canonical at another height in the L1 chain,
+// and the same holds for all its ancestors.
+func FindL2Heads(ctx context.Context, cfg *rollup.Config, l1 L1Chain, l2 L2Chain) (result *FindHeadsResult, err error) {
+	// Fetch current L2 forkchoice state
+	result, err = currentHeads(ctx, cfg, l2)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch current L2 forkchoice state: %w", err)
+	}
 
-	// Loop 1. Walk the L2 chain backwards until we find an L2 block whose L1 origin is canonical.
+	// Remember original unsafe block to determine reorg depth
+	prevUnsafe := result.Unsafe
 
 	// Current L2 block.
-	n := start
-
-	// Number of blocks between n and start.
-	reorgDepth := 0
+	n := result.Unsafe
 
-	// Blockhash of L1 origin hash for the L2 block during the previous iteration, 0 for first
-	// iteration. When this changes as we walk the L2 chain backwards, it means we're seeing a different
-	// (earlier) epoch.
-	var prevL1OriginHash common.Hash
+	var highestL2WithCanonicalL1Origin eth.L2BlockRef
+	var l1Block eth.L1BlockRef
+	var ahead bool
```

Comment:
Comments on what `ahead` means + what the loop invariants / stages are would be helpful